[IDE] fix shift-enter behavior in IDE

brianlheim opened the pull request:
Purpose and Motivation
----------------------

shift-enter needs to be intercepted at the JS level and call a slot on a
C++ object in order to evaluate a line. This shortcut isn't propagated
normally. This commit adds QWebChannel and QWebSockets to the dependencies
and uses them to allow JS to request the help browser C++ code to evaluate
a line.

Also some changes to JS code and SCDoc renderer code

Note that the WebChannel should only be loaded on pages with code to
execute.

The WebChannel is constructed in the IDE's main() and then the help browser
is published to it. On the JS side, evaluateSelection(false) is called
whenever shift-enter on a code snippet is detected.

Note that I based this off @htor's other branch, since that will be
merged soon and since I needed to interface with some of the codemirror code.

Comments on JS style/design welcome, I dunno what I'm doing there.

Fixes #3877.

Types of changes
----------------

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

Checklist
---------

- [x] All tests are passing

- [x] If necessary, new tests were created to address changes in PR (and tests are passing)

- [x] Updated documentation, if necessary

- [x] This PR is ready for review

Remaining Work
------------

- [x] ~May need to update guides since new libs are required (QWebChannel, QWebSockets).~ added qt5xxwebsockets to lists of dependencies in travis and readmes.
- [x] ~Need to check on other platforms (I don't anticipate any problems there, though)~ reportedly works fine on linux.
- [x] ~Update scdoc_version?~ updated
- [x] ~May need to credit code in the new files, which were taken with very minor edits from the Qt WebChannel Standalone Example (https://doc.qt.io/qt-5/qtwebchannel-standalone-example.html)~ included BSD license which is compatible with GPL3.

Also note that QWebChannel may open up an easy/convenient way for us to
translate configuration settings from the IDE to the browser via the
existing Qt signal/slot interface. I see no reason not to do this.

brianlheim commented:
For the actual PR:

https://github.com/brianlheim/supercollider/compare/brianlheim:htor-topic/docs-editor...topic/shift-enter-eval

Right now it's just one commit so also https://github.com/supercollider/supercollider/pull/3960/commits/30edd158bf15231c99d247e5b73dd596fa96b38c works.

After or before @htor's PR is merged I assume it will be on 3.10, at which point I'll rebase this. (Basically, Everything Will Be OK)

htor commented on HelpSource/editor.js:
```diff
@@ -129,7 +129,13 @@ const selectRegion = (options = { flash: true }) => {
 
-/* returns the code selection or line */
+const evalLine = () => {
+    // Call to C++ help browser. `evaluateSelection(false)` evaluates a line. It will execute
```
maybe a little verbose comment, perhaps something like "call c++ evaluation function in order to capture shift-enter" would be fine. the comment should go on top of the function signature.

htor commented on HelpSource/scdoc.js:
```diff
@@ -193,3 +193,35 @@ function fixTOC() {
+    fixTOC();
+
+    // Check that webchannel.js was loaded
+    if(typeof QWebChannel != "undefined") {
```
`!==` is preferred in js instead of `!=` by convention. because less bugs and weird behaviour.

htor commented on editors/sc-ide/core/main.cpp:
```diff
@@ -124,6 +127,22 @@ int main( int argc, char *argv[] )
 
+    // setup HelpBrowser server
+    QWebSocketServer server("SCIDE HelpBrowser Server", QWebSocketServer::NonSecureMode);
+    if (!server.listen(QHostAddress::LocalHost, 12345)) {
```
should we have another port number or is this going to be fine?

brianlheim replied:
A question I have as well!

miczac replied:
pick something in the range of 12323 and 12344 for instance.
based on: 
https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt
12345 should be fine though: LocalHost is apparently hard-coded anyway and 12345 is not in the well-known ports.
Only an overzealous security tool (e.g. on Windows) might be irritated since there are trojans out-there using 12345. (https://www.speedguide.net/port.php?port=12345)

htor commented:
tested on macos and shift-enter works!

but, there are alot of warnings now when i start sc. possibly related to qwebchannel stuff:

```
~/Projects/supercollider/build2 $ ./Install/SuperCollider/SuperCollider.app/Contents/MacOS/SuperCollider 
Property 'modal'' of object 'ScIDE::HelpBrowser' has no notify signal and is not constant, value updates in HTML will be broken!
Property 'windowModality'' of object 'ScIDE::HelpBrowser' has no notify signal and is not constant, value updates in HTML will be broken!
Property 'enabled'' of object 'ScIDE::HelpBrowser' has no notify signal and is not constant, value updates in HTML will be broken!
Property 'geometry'' of object 'ScIDE::HelpBrowser' has no notify signal and is not constant, value updates in HTML will be broken!
Property 'frameGeometry'' of object 'ScIDE::HelpBrowser' has no notify signal and is not constant, value updates in HTML will be broken!
Property 'normalGeometry'' of object 'ScIDE::HelpBrowser' has no notify signal and is not constant, value updates in HTML will be broken!
Property 'x'' of object 'ScIDE::HelpBrowser' has no notify signal and is not constant, value updates in HTML will be broken!
Property 'y'' of object 'ScIDE::HelpBrowser' has no notify signal and is not constant, value updates in HTML will be broken!
Property 'pos'' of object 'ScIDE::HelpBrowser' has no notify signal and is not constant, value updates in HTML will be broken!
Property 'frameSize'' of object 'ScIDE::HelpBrowser' has no notify signal and is not constant, value updates in HTML will be broken!
Property 'size'' of object 'ScIDE::HelpBrowser' has no notify signal and is not constant, value updates in HTML will be broken!
... (many more)
```

brianlheim commented:
@htor Addressed your comments, thanks! I noticed (and forgot about) the warnings. I solved that by creating a new wrapper that will serve as the singular encapsulated point of comm between the help browser and the IDE. Given the discussion we've had around extending the use of this WebChannel and given that it eliminates the warnings, I think that's warranted.

snappizz commented on SCClassLibrary/SCDoc/SCDocRenderer.sc:
```diff
@@ -244,10 +246,10 @@ SCDocHTMLRenderer {
+		// QWebChannel access
+		<< "<script src='qrc:///qtwebchannel/qwebchannel.js' type='text/javascript'></script>\n"
+		<< "</head>\n"
+		<< "<body onload='didLoadRenderedWindow()'>\n";
```
this is more of a problem with the existing code, but `<body onload="...">` is really old hat. it's better to leave `fixTOC` be, and then write this in the JS:

```
$(function () {
    if (typeof QWebChannel !== "undefined") {
        setUpWebChannel();
    }
});
```

brianlheim replied:
OK, where should I write that?

snappizz replied:
it should go in `scdoc.js`, at the top level. `<body onload="fixTOC()">`, and the `fixTOC()` function itself, should stay as is.

htor commented on HelpSource/scdoc.js:
```diff
@@ -193,3 +193,35 @@ function fixTOC() {
+// WebSocket have loaded.
+function setUpWebChannel() {
+    if (location.search != "")
+        var baseUrl = (/[?&]webChannelBaseUrl=([A-Za-z0-9\-:/\.]+)/.exec(location.search)[1]);
```
just drop the if-part all together! it's never going to be used since we hardcoded `ws://localhost:12345` in the server anyway. (it allows overriding the websocket url by setting a query string `?webChannelBaseUrl=...`.

brianlheim replied:
done

brianlheim commented:
I've updated the top PR comment with new info, AFAICT all remaining items are taken care of now.

brianlheim commented:
This should be ready to go!

patrickdupuis commented on editors/sc-ide/CMakeLists.txt:
```diff
@@ -101,6 +112,10 @@ set ( ide_moc_hdr
     widgets/util/color_widget.hpp
     widgets/util/docklet.hpp
     widgets/util/volume_widget.hpp
+    widgets/util/volume_widget.hpp
```
There's a repeated element in this list.

brianlheim replied:
fixed

brianlheim commented:
This is tested and works on Windows
